58 sanctions UI

danieldunderfelt opened the pull request:
Closes #51 

UI updates for sanctioning. This is a milestone PR and the sanctioning feature is not yet complete after merging!

## Features

This PR covers the following features:

- Graphql schema for sanctions
- Sanctions UI for post-inspection
- Detach filtering, paging and sorting logic from reports and create general components for this.
- Integrate table upgrades, improve table even more
- Toggleable sanctioning state for individual sanction rows

## How to run

This PR should run without further setup, but you may need to update removed enum values in the DB. The PR removed the "readied" inspection state and related user relation type. TypeORM cannot handle this update itself, so if you have rows in the DB with the removed values (https://github.com/HSLdevcom/bultti/commit/ba085f0f3f6fbf5066f7d29a5c5123fb69f53388#diff-6d9bc04ef96a88e7758a393bd4c11aa1d7db1231caf15d2aa6d2c8df2970e21a) you need to change them manually.

## Links

Issue: HSLdevcom/bultti#58

BE PR: HSLdevcom/bultti#193

Divinare commented:
Review testing: it works, nice! \o/

Some enhancements/bug fixes that I found while testing (note: not everything has to be fixed in this card!):

* Table pagination not working properly
  * Rows per line option not working
    1) Go to procurementUnit page
    2) Open a vehicleList with 20++ vehicles
    3) try different "rows per page" options. When I have 54 vehicles, isGrouped=false, rows per page 50, page 1 shows only 4 vehicles and page 0 shows 0
  * if using UI pagination, remove table pagination if you have <= 20 rows
    * table pagination: max rows per page: rows.length

* Btw as a personal preference, button texts "Edellinen" and "Seuraava" could be bigger, even 1 rem or that 0.875 is fine for me.

* New task idea: cronjob to preload HFP from inspection time ranges? Then we can extract HFP loading out of POST inspection page

* I was able to create POST inspection before creating PRE inspection, should there be some kind of selection "attach PRE inspection to POST inspection" before starting to fill POST inspection form? Now it requires me to somehow attach PRE inspection to POST inspection

* PRE/POST inspection page: without PRE/POST inspections, the page says nothing, shows just an empty page. Add a notification "Ennakkotarkastuksia ei löytynyt" (because how a user knows if it's loading / bugged / really has no PRE/POST inspections -> card?

* New task idea: pre check & guidance before opening POST inspection form. That guidance would give instructions to the user to setup everything correctly so that he doesn't have to go back and forth between POST inspection page and other pages!
  * Problems that could be avoided by this:
    * Countered a usability problem 1: when creating a pre inspection, I set name, date and departureBlocks. Now it says vehicleList is missing
    * Countered a usability problem 2: when creating a pre inspection, I set name, date and departureBlocks. Now it says contract is missing even though I have it
    * If contracts are invalid, progress should be prevented with note "fix contracts" and after contracts are fixed & valid, user should be able to progress forward

* PRE Inspection: start & end date should be empty by default (user has to select date range)
   * generate date options from season start to season end
   * same bug in POST inspection

* Edit inspectionDates: prevent removing inspectionDates that are already used
  * Say which operator is using and blocking inspectionDates removal

* rename % Osuus -> Suunniteltu % Osuus

* Add (?) hintText after column names with some explanation

* Go back from sanctioning functionality is missing

Some notes here ^, you can decide whether you want to fix them here or make a new card of them!

danieldunderfelt commented:
Tasks belonging to this PR:

- Going back to draft state from sanctioning
- Fix problem where missing contract info crashed the collection of sanctions. It should have used a default value instead. Contract can be validated separately with the Inspection Validation feature.

The rest of the observations and ideas can be made as separate tickets. I made ticket HSLdevcom/bultti#203 for preloading HFP data, that is a very good idea.

> That guidance would give instructions to the user to setup everything correctly so that he doesn't have to go back and forth between POST inspection page and other pages!

This is a good thought, but in real production use everything will be done in the correct order. It's only in dev that we're jumping between the features.

> Countered a usability problem 1: when creating a pre inspection, I set name, date and departureBlocks. Now it says vehicleList is missing

I'm not sure that a "vehicleList" is, do you mean equipment catalogue or execution requirement equipment? The same comment as above applies here too, the user will have done everything in the correct order in production use.

> Countered a usability problem 2: when creating a pre inspection, I set name, date and departureBlocks. Now it says contract is missing even though I have it

Probably just an unrefreshed inspection validation. Contracts are also set up waaaay ahead of making any inspections, I don't see a problem for production.

> If contracts are invalid, progress should be prevented with note "fix contracts" and after contracts are fixed & valid, user should be able to progress forward

Yes, some contract rules should maybe be validated, but it is the user's responsibility to upload correct contracts and check them in the contract editor view. Code that uses rules should have default values to prevent crashes.

> I was able to create POST inspection before creating PRE inspection, should there be some kind of selection "attach PRE inspection to POST inspection" before starting to fill POST inspection form? Now it requires me to somehow attach PRE inspection to POST inspection

The current production pre-inspection for the season will be automatically selected. Draft post-inspections can still be created.

Other observations and ideas that I didn't comment on are good ideas and should be fixed in separate issues.

danieldunderfelt commented:
PR updated!

Fixed (both here and in BE):

- Going back to draft state from sanctioning, with removal of sanction rows and related user-inspection relation type.
- Fix problem where missing contract info crashed the collection of sanctions. It should have used a default value instead.
- Removed sanction button when in sanctioning
- Fixed some localisations

Divinare commented:
Some more testing

* Open pre inspection should imo open to a new tab. Now if you want to check it out, you are forced to leave the page where you are editing POST inspection
* I think pre/post inspection pages have too many togglable containers. I understand toggle in equipment list but not in elements like "Tarkastuksen tiedot" or "Suoritevaatimukset
   --> make suoritevaatimukset without togglable container. Card -> ?
   --> merge 2x Tarkastuksen tiedot + the container with "tarkastuksen nimi" as one untogglable container labeled as "Tarkastuksen tiedot". Card -> ?
* Suoritevaatimukset table: no need "Rivejä per sivu" because you only have 10 rows there
* Suoritevaatimukset table: after clicking "Luo suoritevaatimukset ennakkotarkastuksesta", it still says the same "Luo suoritevaatimukset ennakkotarkastuksesta" even though I don't have to click it. Could text be "Päivitä suoritevaatimukset ennakkotarkastuksesta"?
    * Why is there a button "Poista tarkastuksen suoritevaatimukset"?
* "Aloita sanktiointi" button is clickable even though it says "Joiltakin kohteilta puuttuu suoritevaatimus."
  * Clicking didn't do anything though, so only need to disable the button
* Clicking "Aloita sanktiointi" doesn't do anything even though everything seems ok (no red error texts)
  * I think this might be due to that I currently have 2x POST inspections with the same "tarkastusjakso" and btw both are "versio 0" - is that a bug?
  
  Please make for example a bullet point list with suggested features + links to cards to implement them

danieldunderfelt commented:
> Open pre inspection should imo open to a new tab. Now if you want to check it out, you are forced to leave the page where you are editing POST inspection

The user can middle click all links.

> I think pre/post inspection pages have too many togglable containers. I understand toggle in equipment list but not in elements like "Tarkastuksen tiedot" or "Suoritevaatimukset
--> make suoritevaatimukset without togglable container. Card -> ?
--> merge 2x Tarkastuksen tiedot + the container with "tarkastuksen nimi" as one untogglable container labeled as "Tarkastuksen tiedot". Card -> ?

Sure, if you think it's worthwhile to do. I'm not against it :D

> Suoritevaatimukset table: no need "Rivejä per sivu" because you only have 10 rows there

Let's continue work on the tables in another PR.

> Suoritevaatimukset table: after clicking "Luo suoritevaatimukset ennakkotarkastuksesta", it still says the same "Luo suoritevaatimukset ennakkotarkastuksesta" even though I don't have to click it. Could text be "Päivitä suoritevaatimukset ennakkotarkastuksesta"?

Sure, but the action is the same.

> Why is there a button "Poista tarkastuksen suoritevaatimukset"?

If you want to remove them. You need to recreate them again though. This is probably mostly useful in dev :D

> "Aloita sanktiointi" button is clickable even though it says "Joiltakin kohteilta puuttuu suoritevaatimus." Clicking didn't do anything though, so only need to disable the button

Yes

> Clicking "Aloita sanktiointi" doesn't do anything even though everything seems ok (no red error texts).

It takes a while, be patient. I agree there should probably be a processing step here.

> I think this might be due to that I currently have 2x POST inspections with the same "tarkastusjakso" and btw both are "versio 0" - is that a bug?

No, the version is set when the inspection is published. Then it would take the previously published version and add one. So you can have many inspections in draft

danieldunderfelt commented:
More tasks to do in this PR:

- Disable sanctions button when proceeding is not possible
- Add a processing step while sanctions are collected

Divinare commented:
"The user can middle click all links."

Btw for me that middle click does nothing

danieldunderfelt commented:
Oh, you're right! It should be a real link, but it's probably a button. Yes, in this case I'll try to make it open in a new window bu default.

danieldunderfelt commented:
This can be tested again. Pull both UI and server!

Divinare commented on src/postInspection/SanctionsContainer.tsx:
```diff
@@ -0,0 +1,302 @@
+      id
+      appliedSanctionAmount
+    }
+  }
```
Rename?
```
let updateSanctionsMutation = gql`
  mutation updateSanctions($sanctionUpdates: [SanctionUpdate!]!) {
  ...
```

danieldunderfelt replied:
To what and why? I don't understand this comment.

Divinare replied:
```
let setSanctionMutation = gql`
  mutation setSanction($sanctionUpdates: [SanctionUpdate!]!) {
    updateSanctions(sanctionUpdates: $sanctionUpdates) {
      id
      appliedSanctionAmount
    }
  }
`
```
rename as
```
let updateSanctionsMutation = gql`
  mutation updateSanctions($sanctionUpdates: [SanctionUpdate!]!) {
    updateSanctions(sanctionUpdates: $sanctionUpdates) {
      id
      appliedSanctionAmount
    }
  }
```
because setSanctionMutation and setSanction are way different than the resolver updateSanctions. The latter one is plural and set+update, well they could be the same.

Divinare commented on src/common/table/StatefulTable.tsx:
```diff
@@ -0,0 +1,90 @@
+            setFilters={setFilters}
+            fieldLabels={columnLabels}
+            excludeFields={['id', '__typename']}
+          />
```
There are two ways of implementing this: having a separate StatefulTable or filters / pagination inside Table

-> You have less Table props with StatefulTable but also one additional component which name doesn't indicate much

Not sure which one is the best at the current state but if pagination were made only in UI (backend could have row length limit like 5000), we could put filters inside Table.

danieldunderfelt replied:
StatefulTable is a convenience component for rendering a table based on a paged/filtered/sorted response. Otherwise it would need to be done separately in sanctions and reports, and in future places that work like this.

I'm not sure what your standard for a name is. StatefulTable indicates that the table uses state and works with stateful data. A name is not documentation, it is not expected to describe everything about something. Documentation is forthcoming as soon as we arrive at a good and stable implementation.

Generally I'd like Table to be as dumb as possible and just render the rows given to it. Then we'd have container components like this to compose Table with relevant other components as needed. Sorting MUST be done in the table, but paging could be extracted. But we did agree that it goes in Table for now.

Divinare replied:
It's true that name can't always be descriptive, didn't like it should.

Hmm I like the idea of Table being dumb and simple. What if we had dumb Table inside Table? I mean, merge the current Table.tsx and StatefulTable.tsx as StatefulTable and then extract Table.tsx out (which takes only simple props and renders a table).

danieldunderfelt commented on src/common/components/Tabs.tsx:
```diff
@@ -101,6 +102,7 @@ export type TabChildProps = {
   path?: string
   loading?: boolean
   testId?: string
+  default?: boolean
```
Not for us to decide, Reach Router uses default: https://reach.tech/router/api/RouteComponent

Divinare replied:
Ouch. Well then a comment could be added: default?: boolean // When no children match, a route component with a default prop will render

Btw it seems like it's not using default props

danieldunderfelt commented on src/common/table/Table.tsx:
```diff
@@ -457,37 +481,44 @@ const Table = observer(
     onCancelEdit,
     onSaveEdit,
     pendingValues = [],
+    isAlwaysEditable = false,
```
In normal operation you have to double-click the cell to start editing it, but for sanctions I wanted the editable cell to be, well, always editable, without the need for a double-click.

Documenting the Table component is a task in itself.

danieldunderfelt commented on src/common/table/Table.tsx:
```diff
@@ -457,37 +481,44 @@ const Table = observer(
     setSort: propSetSort,
-  }: PropTypes<ItemType>) => {
+    disablePaging = false,
+    pageState: propPageConfig,
```
Sure, we can use pageState.

danieldunderfelt replied:
Fixed!

danieldunderfelt commented on src/common/table/tableUtils.tsx:
```diff
@@ -0,0 +1,77 @@
+}
+
+// Sensible default rendering function for nice looking data.
+export function useRenderCellValue() {
```
Well, for now. It started as something I used in other components, and it can still be used in other components. So it is a utility. TableRow renders data just as it is by default, a renderCellValue function can be set to customize it.

danieldunderfelt commented on src/common/table/Table.tsx:
```diff
@@ -457,37 +481,44 @@ const Table = observer(
+    pageState: propPageConfig,
+    setPage: propSetPage,
+    setPageSize: propSetPageSize,
+    pageMeta: propPageMeta,
```
PageMeta is the data that describes how the response data is paged. If it's not paged, the Table creates internal pageMeta which is passed to the paging component for UI paging.

I'm not sure what you propose but I am all for taking the Paging stuff out from Table.

danieldunderfelt commented on src/common/table/Table.tsx:
```diff
@@ -187,144 +187,154 @@ export const CellContent = styled.div<{ footerCell?: boolean }>`
   sort?: SortConfig[]
   setSort?: (arg: ((sort: SortConfig[]) => SortConfig[]) | SortConfig[]) => unknown
-}
+  disablePaging?: boolean
```
If no paging stuff is supplied to Table, it will do UI paging internally. With this, UI paging can be disabled.

danieldunderfelt commented:
Fixed pro and var naming related to page things.

Divinare reviewed (approved):
Approving at this point since this work continues in other PR's!
